### PR TITLE
python unit tests: adds more tests for flow, generators and engineering

### PR DESCRIPTION
**Pull request by PetholzA**
- test_flow should cover all existing functions
- generators is still in progress
- engineering is still in progress

### CODE REVIEW

**Comment by angriman** [comment]
Thanks for adding more tests. Since this is still ongoing work, I converted it to draft.

**Comment by fabratu** [comment]
ToDo Notes (based on an offline discussion):

- finish generators >=80%
- get engineering to 100% (basic functions, used everywhere)

File: networkit/test/test_generators.py [hunk 1]
```diff
@@ -13,6 +16,44 @@ def testClusteredRandomGraphGenerator(self):
...
         self.assertLessEqual(nCommunities, k)
+      
+    def testBarabasiAlbertGenerator(self):
+        g = nk.generators.BarabasiAlbertGenerator(2, 50, 1)
+        self.assertEqual(g.generate().numberOfNodes(), 50)
+        self.assertEqual(g.generate().numberOfEdges(), 98)
+
+    def testBarabasiAlbertGeneratorFit(self):
+        g = nk.generators.BarabasiAlbertGenerator(2, 50, 1)
+        g.fit(self.G)
```
Comment (fabratu) [inline networkit/test/test_generators.py:27]: Add asserts to test whether the new graph has the expected properties (e.g. `#nodes`, `#edges`).

File: networkit/test/test_generators.py [hunk 2]
```diff
@@ -13,6 +16,44 @@ def testClusteredRandomGraphGenerator(self):
...
+        self.assertEqual(g.generate().numberOfEdges(), 98)
+
+    def testBarabasiAlbertGeneratorFit(self):
+        g = nk.generators.BarabasiAlbertGenerator(2, 50, 1)
+        g.fit(self.G)  
+  
+    def testBTERReplicator(self):
+        g = nk.generators.BTERReplicator(self.G)
+        g.setPaths("./BTER.data")
+        g.fit(self.G)
```
Comment (fabratu) [inline networkit/test/test_generators.py:32]: Same as above (`fit` function).

File: networkit/test/test_generators.py [hunk 3]
```diff
@@ -13,6 +16,44 @@ def testClusteredRandomGraphGenerator(self):
...
+    
+    def testChungLuGenerator(self):
+        g = nk.generators.ChungLuGenerator([1.3, 1.5, 1.7])
+        self.assertEqual(g.generate().numberOfNodes(), 3)    
+
+    def testChungLuGeneratorFit(self):
+        g = nk.generators.ChungLuGenerator([1.3, 1.5, 1.7])
+        g.fit(self.G)
```
Comment (fabratu) [inline networkit/test/test_generators.py:40]: Same as above (`fit` function).

File: networkit/test/test_generators.py [hunk 4]
```diff
@@ -13,6 +16,44 @@ def testClusteredRandomGraphGenerator(self):
...
+
+    def testDorogovtsevMendesGenerator(self):
+        g = nk.generators.DorogovtsevMendesGenerator(100)
+        self.assertEqual(g.generate().numberOfNodes(), 100)
+
+    def testDorogovtsevMendesGeneratorFit(self):
+        g = nk.generators.DorogovtsevMendesGenerator(100)
+        g.fit(self.G)
```
Comment (fabratu) [inline networkit/test/test_generators.py:48]: Same as above (`fit` function).

File: networkit/test/test_generators.py [hunk 5]
```diff
@@ -50,7 +96,86 @@ def testEdgeSwitchingMarkovChainGenerator(self):
                 self.assertEqual(G.degree(u), degseq[u])
 
             break
+
+    """def testEdgeSwitchingMarkovChainGeneratorFit(self):
```
Comment (fabratu) [inline networkit/test/test_generators.py:100]: Is this intended?

File: networkit/test/test_generators.py [hunk 6]
```diff
@@ -50,7 +96,86 @@ def testEdgeSwitchingMarkovChainGenerator(self):
...
+        g = nk.generators.EdgeSwitchingMarkovChainGenerator(degseq, False, 10)  
+        g.fit(self.G)"""
+
+    def testErdosRenyiGenerator(self):
+        g = nk.generators.ErdosRenyiGenerator(10, 0.15, True)
+        self.assertEqual(g.generate().numberOfNodes(), 10)
+
+    def testErdosRenyiGeneratorFit(self):
+        g = nk.generators.ErdosRenyiGenerator(10, 0.15, True)
+        g.fit(self.G)
```
Comment (fabratu) [inline networkit/test/test_generators.py:111]: Same as above (`fit` function).

File: networkit/test/test_generators.py [hunk 7]
```diff
@@ -50,7 +96,86 @@ def testEdgeSwitchingMarkovChainGenerator(self):
...
+
+    def testHavelHakimiGenerator(self):
+        g = nk.generators.HavelHakimiGenerator([2,2,2,2])
+        self.assertTrue(g.isRealizable()) 
+        self.assertTrue(g.getRealizable())
+
+    def testHavelHakimiGeneratorFit(self):
+        g = nk.generators.HavelHakimiGenerator([2,2,2,2])  
+        g.fit(self.G)
```
Comment (fabratu) [inline networkit/test/test_generators.py:120]: Same as above (`fit` function).

File: networkit/test/test_generators.py [hunk 8]
```diff
@@ -50,7 +96,86 @@ def testEdgeSwitchingMarkovChainGenerator(self):
...
+        g.run()
+        self.assertEqual(g.getGraph().numberOfEdges(), 0)
+        self.assertEqual(g.getPartition().numberOfElements(), 5)
+
+    def testLFRGeneratorFit(self):
+        g = nk.generators.LFRGenerator(5)
+        g.setDegreeSequence([0,0,0,0,0])
+        g.setMu(1.5)
+        g.setCommunitySizeSequence([1,1,1,1,1])
+        g.fit(self.G)
```
Comment (fabratu) [inline networkit/test/test_generators.py:144]: Same as above (`fit` function).

File: networkit/test/test_generators.py [hunk 9]
```diff
@@ -50,7 +96,86 @@ def testEdgeSwitchingMarkovChainGenerator(self):
...
+        g2.run()
+        self.assertEqual(g2.getGraph().numberOfNodes(), 100)
+        self.assertEqual(g2.getPartition().numberOfElements(), 100)
+
+    def testLFRGeneratorPowerlawDegreeSequenceFit(self):
+        g2 = nk.generators.LFRGenerator(10)
+        g2.generatePowerlawDegreeSequence(5, 6, -2)
+        g2.generatePowerlawCommunitySizeSequence(5, 6, -1)
+        g2.setMu(0.5)
+        g2.fit(self.G, vanilla=True, plfit=False)
```
Comment (fabratu) [inline networkit/test/test_generators.py:162]: Same as above (`fit` function). Also, better split the fit-test for `LFRGenerator` if different parameters use different lines of code.

File: networkit/test/test_generators.py [hunk 10]
```diff
@@ -67,6 +192,21 @@ def testPubWebGenerator(self):
...
+        self.assertAlmostEqual(resGraph.numberOfNodes(), 4)
+        self.assertAlmostEqual(resGraph.numberOfEdges(), 4)
+
+    def testRmatGeneratorFit(self):
+        g = nk.generators.RmatGenerator(2, 1, 0.25, 0.25, 0.25, 0.25)
+        g.setPaths("./rMat.data")
+        inputGraph = nk.Graph(3)
+        inputGraph.addEdge(0,1)
+        inputGraph.addEdge(1,2)
+        g.fit(inputGraph, kronfit=False)
```
Comment (fabratu) [inline networkit/test/test_generators.py:208]: Same as above (`fit` function).

**Review by fabratu — CHANGES_REQUESTED** [review]

File: networkit/test/test_generators.py [hunk 11]
```diff
@@ -13,6 +16,38 @@ def testClusteredRandomGraphGenerator(self):
...
+        g = nk.generators.BarabasiAlbertGenerator(2, 50, 1)
+        g.fit(self.G)  
+        self.assertEqual(g.generate().numberOfNodes(), 50)
+        self.assertEqual(g.generate().numberOfEdges(), 98)
+  
+    def testBTERReplicator(self):
+        g = nk.generators.BTERReplicator(self.G)
+        g.setPaths("./BTER.data")
+        g.fit(self.G)
+        #print(g.generate().numberOfNodes())
```
Comment (fabratu) [inline networkit/test/test_generators.py:30]: Remove unnecessary lines.

**Review by fabratu — APPROVED** [review]

**Pull request merged**